Test TestInstanceAutoJoin.testAutoRegistrationShouldFailWhenWaitingResponse is flaky

- Flaky
LOG:
>020-08-12T12:38:16.1190369Z [ERROR] testAutoRegistrationShouldFailWhenWaitingResponse(org.apache.helix.integration.paticipant.TestInstanceAutoJoin)  Time elapsed: 0.439 s  <<< FAILURE!
2020-08-12T12:38:16.1194078Z java.lang.AssertionError: expected:<true> but was:<false>
2020-08-12T12:38:16.1202161Z 	at org.apache.helix.integration.paticipant.TestInstanceAutoJoin.testAutoRegistrationShouldFailWhenWaitingResponse(TestInstanceAutoJoin.java:113)

>-08-08T10:12:24.8433385Z [ERROR] testAutoRegistrationShouldFailWhenWaitingResponse(org.apache.helix.integration.paticipant.TestInstanceAutoJoin)  Time elapsed: 0.366 s  <<< FAILURE!
2020-08-08T10:12:24.8433664Z java.lang.AssertionError: expected:<true> but was:<false>
2020-08-08T10:12:24.8434622Z 	at org.apache.helix.integration.paticipant.TestInstanceAutoJoin.testAutoRegistrationShouldFailWhenWaitingResponse(TestInstanceAutoJoin.java:113)

Theory and proof:
Running in local environment, the code would always work. With exception like:
>248277 [ClusterManager_Watcher_CLUSTER_TestInstanceAutoJoin_localhost_279700_PARTICIPANT_10008] ERROR org.apache.helix.manager.zk.ZKHelixManager  - fail to createClient.
org.apache.helix.HelixException: Failed to get Azure cloud instance information from url {}
	at org.apache.helix.cloud.azure.AzureCloudInstanceInformationProcessor.getAzureCloudInformationFromUrl(AzureCloudInstanceInformationProcessor.java:122)
	at org.apache.helix.cloud.azure.AzureCloudInstanceInformationProcessor.fetchCloudInstanceInformation(AzureCloudInstanceInformationProcessor.java:95)
	at org.apache.helix.manager.zk.ParticipantManager.getCloudInstanceInformation(ParticipantManager.java:225)
	at org.apache.helix.manager.zk.ParticipantManager.joinCluster(ParticipantManager.java:201)
	at org.apache.helix.manager.zk.ParticipantManager.handleNewSession(ParticipantManager.java:141)
	at org.apache.helix.manager.zk.ZKHelixManager.handleNewSessionAsParticipant(ZKHelixManager.java:1359)
	at org.apache.helix.manager.zk.ZKHelixManager.handleNewSession(ZKHelixManager.java:1319)
	at org.apache.helix.manager.zk.ZKHelixManager.createClient(ZKHelixManager.java:749)
	at org.apache.helix.manager.zk.ZKHelixManager.connect(ZKHelixManager.java:786)
	at org.apache.helix.integration.manager.MockParticipantManager.run(MockParticipantManager.java:98)
	at java.lang.Thread.run(Thread.java:745)

This exception is due to retrieving the cloud config would fail in local environment from 169.254.169.254 configured for cloud. This exception is a must for the test to succeed because if prevent ParticipantManager from creating liveInstance after autojoin().

If the run in GitHub testing machine running this unit test would be able to access `169.254.169.254`. This exception may not throw and thus the test may fail.

Search the log from failed test and we did not see this exception log. 
https://pipelines.actions.githubusercontent.com/sLVPVVo6suizPM5gIrHDigmaxapcQ7bGBvxdUFsAfGjAfNkqWG/_apis/pipelines/1/runs/188/signedlogcontent/3?urlExpires=2020-08-12T19%3A51%3A15.4635058Z&urlSigningMethod=HMACV1&urlSignature=5Z2u1uaZZHr3ScEm6sY7qj%2F77j%2FR4ArPftTUxejGaQg%3D

Thus, a strong support to the theory we need configure for sure to fail endpoint to retrieve cloud information.

```
   public void testAutoRegistrationShouldFailWhenWaitingResponse() throws Exception {
    // Create CloudConfig object and add to config
    CloudConfig.Builder cloudConfigBuilder = new CloudConfig.Builder();
    cloudConfigBuilder.setCloudEnabled(true);
    cloudConfigBuilder.setCloudProvider(CloudProvider.AZURE);
    cloudConfigBuilder.setCloudID("TestID");
    CloudConfig cloudConfig = cloudConfigBuilder.build();

    HelixManager manager = _participants[0];
    HelixDataAccessor accessor = manager.getHelixDataAccessor();

    _gSetupTool.addResourceToCluster(CLUSTER_NAME, db3, 60, "OnlineOffline",
        RebalanceMode.FULL_AUTO + "");
    _gSetupTool.rebalanceStorageCluster(CLUSTER_NAME, db3, 1);
    String instance3 = "localhost_279700";

    ConfigScope scope = new ConfigScopeBuilder().forCluster(CLUSTER_NAME).build();

    manager.getConfigAccessor().set(scope, ZKHelixManager.ALLOW_PARTICIPANT_AUTO_JOIN, "true");
    // Write the CloudConfig to Zookeeper
    PropertyKey.Builder keyBuilder = accessor.keyBuilder();
    accessor.setProperty(keyBuilder.cloudConfig(), cloudConfig);

    MockParticipantManager autoParticipant =
        new MockParticipantManager(ZK_ADDR, CLUSTER_NAME, instance3);
    autoParticipant.syncStart();

    Assert.assertTrue(null == manager.getHelixDataAccessor()                  -----------> failure
        .getProperty(accessor.keyBuilder().liveInstance(instance3)));
    try {
      manager.getConfigAccessor().getInstanceConfig(CLUSTER_NAME, instance3);
      fail(
          "Exception should be thrown because the instance cannot be added to the cluster due to the disconnection with Azure endpoint");
    } catch (HelixException e) {

    }
  }
```

- leak of one participant manager, minor

--- a/helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInstanceInformationProcessor.java
+++ b/helix-core/src/main/java/org/apache/helix/cloud/azure/AzureCloudInstanceInformationProcessor.java
@@ -109,7 +109,7 @@ public class AzureCloudInstanceInformationProcessor
       CloseableHttpResponse response = _closeableHttpClient.execute(httpGet);
       if (response == null || response.getStatusLine().getStatusCode() != 200) {
         String errorMsg = String.format(
-            "Failed to get an HTTP Response for the request. Response: {}. Status code: {}",
+            "Failed to get an HTTP Response for the request. Response: %s. Status code: %s",
             (response == null ? "NULL" : response.getStatusLine().getReasonPhrase()),
             response.getStatusLine().getStatusCode());
         throw new HelixException(errorMsg);
@@ -119,7 +119,7 @@ public class AzureCloudInstanceInformationProcessor
       return responseString;
     } catch (IOException e) {
       throw new HelixException(
-          String.format("Failed to get Azure cloud instance information from url {}", url), e);
+          String.format("Failed to get Azure cloud instance information from url %s", url), e);
     }
   }
 
@@ -153,7 +153,7 @@ public class AzureCloudInstanceInformationProcessor
       }
     } catch (IOException e) {
       throw new HelixException(
-          String.format("Error in parsing cloud instance information: {}", response, e));
+          String.format("Error in parsing cloud instance information: %s", response, e));
     }
     return azureCloudInstanceInformation;
   }

--- a/helix-core/src/test/java/org/apache/helix/integration/paticipant/TestInstanceAutoJoin.java
+++ b/helix-core/src/test/java/org/apache/helix/integration/paticipant/TestInstanceAutoJoin.java
@@ -4,6 +4,7 @@ import org.apache.helix.HelixDataAccessor;
 import org.apache.helix.HelixException;
 import org.apache.helix.HelixManager;
 import org.apache.helix.PropertyKey;
+import org.apache.helix.TestHelper;
 import org.apache.helix.cloud.constants.CloudProvider;
 import org.apache.helix.integration.common.ZkStandAloneCMTestBase;
 import org.apache.helix.integration.manager.MockParticipantManager;
@@ -14,8 +15,6 @@ import org.apache.helix.model.IdealState.RebalanceMode;
 import org.apache.helix.model.builder.ConfigScopeBuilder;
 import org.testng.Assert;
 import org.testng.annotations.Test;
-
-import static org.testng.Assert.*;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one
@@ -83,12 +82,11 @@ public class TestInstanceAutoJoin extends ZkStandAloneCMTestBase {
   }
 
   @Test(dependsOnMethods = "testInstanceAutoJoin")
-  public void testAutoRegistrationShouldFailWhenWaitingResponse() throws Exception {
+  public void testAutoRegistration() throws Exception {
     // Create CloudConfig object and add to config
     CloudConfig.Builder cloudConfigBuilder = new CloudConfig.Builder();
     cloudConfigBuilder.setCloudEnabled(true);
     cloudConfigBuilder.setCloudProvider(CloudProvider.AZURE);
-    cloudConfigBuilder.setCloudID("TestID");
     CloudConfig cloudConfig = cloudConfigBuilder.build();
 
     HelixManager manager = _participants[0];
@@ -110,14 +108,23 @@ public class TestInstanceAutoJoin extends ZkStandAloneCMTestBase {
         new MockParticipantManager(ZK_ADDR, CLUSTER_NAME, instance3);
     autoParticipant.syncStart();
 
-    Assert.assertTrue(null == manager.getHelixDataAccessor()
-        .getProperty(accessor.keyBuilder().liveInstance(instance3)));
+    // if the test is run in cloud environment, auto registration will succeed and live instance
+    // will be added, otherwise, auto registration will fail and instance config will not be
+    // populated. An exception will be thrown.
     try {
       manager.getConfigAccessor().getInstanceConfig(CLUSTER_NAME, instance3);
-      fail(
-          "Exception should be thrown because the instance cannot be added to the cluster due to the disconnection with Azure endpoint");
+      Assert.assertTrue(TestHelper.verify(() -> {
+        if (null == manager.getHelixDataAccessor()
+            .getProperty(accessor.keyBuilder().liveInstance(instance3))) {
+          return false;
+        }
+        return true;
+      }, 2000));
     } catch (HelixException e) {
-
+      Assert.assertTrue(null == manager.getHelixDataAccessor()
+          .getProperty(accessor.keyBuilder().liveInstance(instance3)));
     }
+
+    autoParticipant.syncStop();
   }
 }

